checker/src/main/java/alluxio/checker/MapReduceIntegrationChecker.java

```diff
@@ -0,0 +1,353 @@
+    System.setProperty("hadoop.home.dir", "/");
+    sFileSystem = FileSystem.get(URI.create(conf.get("fs.defaultFS")), conf);
+
+    String stringPath = "/user/" + systemUserName + "/";
+    Path userPath = new Path(stringPath);
+
+    // Creates user folder if not exists
+    if (!sFileSystem.exists(userPath)) {
+      sFileSystem.mkdirs(userPath);
+      LOG.info("User Path " + stringPath + " created.");
+    }
+    sInputFilePath = new Path(userPath + "/MapReduceInputFile");
+    sOutputFilePath = new Path(userPath + "/MapReduceOutputFile");
+  }
+
+  /**
+   * Creates the MapReduce input file and delete previous output file if exists.
+   */
+  private static void createInputFile(int inputSplits) throws Exception {
+    if (!sFileSystem.exists(new Path(sInputFilePath + "/1.txt"))) {
+      LOG.info("Begin Write MapReduce input file into hdfs");
+      for (int i = 1; i <= inputSplits; i++) {
+        FSDataOutputStream inputFileStream = sFileSystem
+            .create(new Path(sInputFilePath + "/" + i + ".txt"));
+        inputFileStream.writeByte(1);
+        inputFileStream.close();
+      }
+      LOG.info("End Write MapReduce input file into hdfs");
+    }
+
+    if (sFileSystem.exists(sOutputFilePath)) {
+      sFileSystem.delete(sOutputFilePath, true);
+    }
+  }
+
+  /**
+   * @return whether MapReduce can find Alluxio classes and filesystem
+   */
+  private static boolean generateReport() throws Exception {
+    // if we do not have output file or MapReduce job is not success, we do not generate report
+    if (!sFileSystem.exists(sOutputFilePath)) {
+      return false;
+    }
+    if (!sFileSystem.exists(new Path(sOutputFilePath + "/_SUCCESS"))) {
+      return false;
+    }
+
+    // Reads all the part-r-* files in MapReduceOutPutFile folder
+    sFileSystem.delete(new Path(sOutputFilePath + "/_SUCCESS"), true);
+    FileStatus[] outputFileStatus = sFileSystem.listStatus(sOutputFilePath);
+
+    boolean canFindClass = true;
+    boolean canFindFS = true;
+
+    StringBuilder cannotFindClassBuilder = new StringBuilder();
+    StringBuilder cannotFindFSBuilder = new StringBuilder();
+    StringBuilder successBuilder = new StringBuilder();
+
+    for (int i = 0; i < outputFileStatus.length; i++) {
+      Path curOutputFilePath = outputFileStatus[i].getPath();
+      BufferedReader curOutputFileReader = new BufferedReader(
+          new InputStreamReader(sFileSystem.open(curOutputFilePath)));
+      try {
+        String nextLine = "";
+        while ((nextLine = curOutputFileReader.readLine()) != null) {
+          Status curOutputStatus = Status.valueOf(nextLine
+              .substring(0, nextLine.indexOf(' ')).trim());
+          String curOutputAddresses = nextLine.substring(nextLine.indexOf(' ') + 1).trim();
+          switch (curOutputStatus) {
+            case FAIL_TO_FIND_CLASS:
+              canFindClass = false;
+              cannotFindClassBuilder.append(" ").append(curOutputAddresses);
```

Comment:
after the reduce step, will we ever have more than one output line per failure type?

Reply:
the merged IP addresses of each failure type, no matter its length, will appear in one output line. At least when I check four machines, they are in one output line. I will check more machines.